File: src/core/validators/CommonSimpleType/CommonDecimalDecimalPlacesMustNotBeGreaterThanTotalDigits.ts
import {ValidationRuleBase} from "../ValidationRuleBase";

var MetaEdGrammar = require("../../../../src/grammar/gen/MetaEdGrammar").MetaEdGrammar;

export class CommonDecimalDecimalPlacesMustNotBeGreaterThanTotalDigits extends ValidationRuleBase
{
    public handlesContext(context: any) : boolean {
        return context.ruleIndex === MetaEdGrammar.RULE_commonDecimal;
    }

    public isValid(context: any): boolean {
        let decimalPlaces = context.decimalPlaces().UNSIGNED_INT().getText();
        let totalDigits = context.totalDigits().UNSIGNED_INT().getText();
        return Number(decimalPlaces) <= Number(totalDigits);
    }

    public getFailureMessage(context: any): string {
        return `Common Decimal '${context.commonDecimalName().getText()} has decimal places greater than total digits.`;
    }
}

File: src/core/validators/CommonSimpleType/CommonDecimalDecimalPlacesMustNotBeGreaterThanTotalDigits.cs
using System;
using MetaEd.Grammar.Antlr;

namespace MetaEd.Core.Validator.CommonSimpleType
{
    public class CommonDecimalDecimalPlacesMustNotBeGreaterThanTotalDigits : ValidationRuleBase<MetaEdGrammar.CommonDecimalContext>
    {
        public override bool IsValid(MetaEdGrammar.CommonDecimalContext context)
        {
            // if there are convert exceptions, let it bomb out -- language parser should have handled
            var decimalPlaces = context.decimalPlaces().DecimalPlaces();
            var totalDigits = context.totalDigits().TotalDigits();

            return Convert.ToInt32(decimalPlaces) <= Convert.ToInt32(totalDigits);
        }

        public override string GetFailureMessage(MetaEdGrammar.CommonDecimalContext context)
        {
            return string.Format("Common Decimal '{0}' has decimal places greater than total digits.", context.commonDecimalName().GetText());
        }
    }
}
